lib/cli.js

```diff
@@ -278,26 +279,36 @@ exports.process = () => {
 }
 
 exports.run = () => {
-  const config = exports.process()
-
-  switch (config.cmd) {
-    case 'start':
-      new Server(config).start()
-      break
-    case 'run':
-      require('./runner')
-        .run(config)
-        .on('progress', printRunnerProgress)
-      break
-    case 'stop':
-      require('./stopper').stop(config)
-      break
-    case 'init':
-      require('./init').init(config)
-      break
-    case 'completion':
-      require('./completion').completion(config)
-      break
+  const cliOptions = exports.process()
+  const cmd = cliOptions.cmd // prevent config from changing the command
+  const cmdSupportsConfigPromise =
+    cmd === 'start' || cmd === 'run' || cmd === 'stop'
+  if (!cmdSupportsConfigPromise) {
+    switch (cmd) {
+      case 'init':
+        require('./init').init(cliOptions)
+        break
+      case 'completion':
+        require('./completion').completion(cliOptions)
+        break
+    }
+  } else {
+    cfg.parseConfig(cliOptions.configFile, cliOptions, { promiseConfig: true })
```

Comment:
Let's set `throwErrors: true`, add a `.catch()` and `process.exit(1)` in the catch block. The idea is that `throwErrors: false` is deprecated and we want to remove it to cleanup the parseConfig() implementation in the next major. Hence we should not rely onto the deprecated behaviour ourselves.

Reply:
Are we catching just the parseConfig? or the entire switch?

<details>
<summary>Option 1: Catch Parse Config and All Commands</summary>

```js
exports.run = () => {
  const cliOptions = exports.process()
  const cmd = cliOptions.cmd // prevent config from changing the command
  const cmdNeedsConfig = cmd === 'start' || cmd === 'run' || cmd === 'stop'
  function failCommand (err) {
    console.error(err) // TODO: configure a CLI Logger?

    // The `run` function is a private application, not a public API. We don't
    // need to worry about process.exit vs throw vs promise rejection here.
    process.exit(1)
  }
  if (cmdNeedsConfig) {
    cfg.parseConfig(
      cliOptions.configFile,
      cliOptions,
      {
        promiseConfig: true,
        throwErrors: true
      }
    ).then(function onKarmaConfigFulfilled (config) {
      switch (cmd) {
        case 'start':
          new Server(config).start()
          break
        case 'run':
          require('./runner')
            .run(config)
            .on('progress', printRunnerProgress)
          break
        case 'stop':
          require('./stopper').stop(config)
          break
      }
    }).catch(function onCommandFailed (reason) {
      return failCommand(reason)
    })
  } else {
    try {
      switch (cmd) {
        case 'init':
          require('./init').init(cliOptions)
          break
        case 'completion':
          require('./completion').completion(cliOptions)
          break
      }
    } catch (commandException) {
      return failCommand(commandException)
    }
  }
}
```

</details>

<details>
<summary>Option 2: Catch Parse Config Only</summary>

```js
exports.run = () => {
  const cliOptions = exports.process()
  const cmd = cliOptions.cmd // prevent config from changing the command
  const cmdNeedsConfig = cmd === 'start' || cmd === 'run' || cmd === 'stop'
  if (cmdNeedsConfig) {
    cfg.parseConfig(
      cliOptions.configFile,
      cliOptions,
      {
        promiseConfig: true,
        throwErrors: true
      }
    ).then(function onKarmaConfigFulfilled (config) {
      switch (cmd) {
        case 'start':
          new Server(config).start()
          break
        case 'run':
          require('./runner')
            .run(config)
            .on('progress', printRunnerProgress)
          break
        case 'stop':
          require('./stopper').stop(config)
          break
      }
    }, function onKarmaConfigRejected (reason) {
      console.error(reason) // TODO: configure a CLI Logger?

      // The `run` function is a private application, not a public API. We don't
      // need to worry about process.exit vs throw vs promise rejection here.
      process.exit(1)
    })
  } else {
    switch (cmd) {
      case 'init':
        require('./init').init(cliOptions)
        break
      case 'completion':
        require('./completion').completion(cliOptions)
        break
    }
  }
}
```

</details>

Reply:
Option 2.

> console.error(reason) // TODO: configure a CLI Logger?

Oh, now I remember why I've added item (4) to the list! Because we need to call `setupFromConfig()` to configure logger before we actually attempt to parse config. So we should call it here and guard calls in server/runner/stopper by the check for the backward compatibility.

And we should also expose it on the public API for API consumers. Maybe under a different name. I'll get back on this part.

Reply:
I actually came up with a better idea. When `promiseConfig == true`, make a call to `setupFromConfig()` at the beginning of `parseConfig()` function. This way we still get logs, if something fails during config parsing and we don't need to expose any new public API.

And guard calls in server/runner/stopper by the `instanceof Config` check for the backward compatibility. They will be cleaned up in the next major together with the other old parsing logic.

Reply:
I'm leaning towards Option 2 or Option 4

What do you think?

<details>
<summary>Option 1: What?</summary>

I don't know exactly how log4js works, so I'm not 100% sold on what I am about to suggest, but I took your previous comment and thought of the following:

Instead of configuring it at the start of the `parseConfig` function, why not configure it at the start of the `fail` function, since that is the point that we know that the other logger setup code won't be run.

```js
// This guarding may not be needed. It doesn't look like the edge case is
// possible with the _current_ code.
let loggerHasBeenSetup = false

/**
 * Allow future calls to `parseConfig` to configure the logger based on a
 * successfully created `Config` instance.
 */
function resetParseConfigLogger () {
  loggerHasBeenSetup = false
}

/**
 * Setup up the logger once in the life time of `parseConfig`. If options aren't
 * provided, it uses the same defaults as the a `Config` class instance.
 *
 * This prevents the edge case of the logger being set up at the end of
 * `finalizeConfig` but before `normalizeConfig`, and then an error occurs
 * which, if interpreted as a critical failure, would then try to configure the
 * logger again within the `fail` function.
 */
function setupLoggerInParseConfig (loggerOptions) {
  if (!loggerHasBeenSetup) {
    if (!loggerOptions) {
      loggerOptions = {}
    }

    logger.setupFromConfig(
      {
        colors: loggerOptions.colors,
        logLevel: loggerOptions.logLevel
      },
      loggerOptions.appenders
    )
    loggerHasBeenSetup = true
  }
}
```

```diff

 function parseConfig (configFilePath, cliOptions, parseOptions) {
   const promiseConfig = parseOptions && parseOptions.promiseConfig === true
   const throwErrors = parseOptions && parseOptions.throwErrors === true
   function fail () {
+    setupLoggerInParseConfig()
     log.error(...arguments)
+
+    // This execution is about to end, so enable future executions to configure
+    // the logger.
+    resetParseConfigLogger()
     // ...

   }
   // ...
   
   function finalizeConfig (config) {
     // ...

     // configure the logger as soon as we can
-    logger.setup(config.logLevel, config.colors, config.loggers)
+    setupLoggerInParseConfig({
+      appenders: config.loggers,
+      colors: config.colors,
+      logLevel: config.logLevel
+    })

     log.debug(configFilePath ? `Loading config ${configFilePath}` : 'No config file specified.')

     return normalizeConfig(config, configFilePath)
   }
   // ...
 }
```

A successfully set up config prevents the logger from being configured again unless parse config is called again and then it encounters an error. This may be a breaking. I wrote it and I find it confusing.

The fact that I don't 100% remember what I was going for isn't a good sign for this option. Maybe the coffee hadn't set in yet.

---

</details>

<details>
<summary>Option 2: Configure logger in <code>fail()</code></summary>

```diff
   function fail () {
+    // `setupFromConfig` provides defaults for `colors` and `logLevel`.
+    // `setup` provides defaults for `appenders`
+    // The first argument MUST BE an object
+    logger.setupFromConfig({})
     log.error(...arguments)
     if (throwErrors) {
```

---

</details>


<details>
<summary>Option 3: When <code>promiseConfig == true</code> OR <code>throwErrors == true</code>, make a call to <code>setupFromConfig()</code> at the beginning of <code>parseConfig()</code> function.</summary>

The thought here being that non-promise failures will want a configured logger too.

```diff
 function parseConfig (configFilePath, cliOptions, parseOptions) {
   const promiseConfig = parseOptions && parseOptions.promiseConfig === true
   const throwErrors = parseOptions && parseOptions.throwErrors === true
+  const shouldSetupLoggerEarly = promiseConfig || throwErrors
+  if (shouldSetupLoggerEarly) {
+    // `setupFromConfig` provides defaults for `colors` and `logLevel`.
+    // `setup` provides defaults for `appenders`
+    // The first argument MUST BE an object
+    logger.setupFromConfig({})
+  }
   function fail () {
```

---

</details>


<details>
<summary>Option 4: When <code>promiseConfig == true</code>, make a call to <code>setupFromConfig()</code> at the beginning of <code>parseConfig()</code> function.</summary>

A first pass at what you mentioned above without modification.

```diff
 function parseConfig (configFilePath, cliOptions, parseOptions) {
   const promiseConfig = parseOptions && parseOptions.promiseConfig === true
   const throwErrors = parseOptions && parseOptions.throwErrors === true
+  const shouldSetupLoggerEarly = promiseConfig
+  if (shouldSetupLoggerEarly) {
+    // `setupFromConfig` provides defaults for `colors` and `logLevel`.
+    // `setup` provides defaults for `appenders`
+    // The first argument MUST BE an object
+    logger.setupFromConfig({})
+  }
   function fail () {
```

---

</details>

Reply:
Both 2 and 4 look good to me. Let's go with the 4 as it is nicer from a backward compatibility persepective.

Reply:
> And guard calls in server/runner/stopper by the instanceof Config check for the backward compatibility. They will be cleaned up in the next major together with the rest of the old parsing logic.

Can you also do this please? Just move them into if/else we alreayd have for `parseConfig()`.

Reply:
Just to make sure I understand correctly, you are looking for something like this in each file:


```diff
 class Server extends KarmaEventEmitter {
   constructor (cliOptionsOrConfig, done) {
     super()
     cliOptionsOrConfig = cliOptionsOrConfig || {}
-    logger.setupFromConfig({
-      colors: cliOptionsOrConfig.colors,
-      logLevel: cliOptionsOrConfig.logLevel
-    })
     this.log = logger.create('karma-server')
     done = helper.isFunction(done) ? done : process.exit
     this.loadErrors = []
 
     let config
     if (cliOptionsOrConfig instanceof cfg.Config) {
       config = cliOptionsOrConfig
     } else {
+      logger.setupFromConfig({
+        colors: cliOptionsOrConfig.colors,
+        logLevel: cliOptionsOrConfig.logLevel
+      })
       const deprecatedCliOptionsMessage =
         'Passing raw CLI options to `new Server(config, done)` is ' +
         'deprecated. Use ' +
         '`parseConfig(configFilePath, cliOptions, {promiseConfig: true, throwErrors: true})` ' +
         'to prepare a processed `Config` instance and pass that as the ' +
         '`config` argument instead.'
       this.log.warn(deprecatedCliOptionsMessage)
       try {
         config = cfg.parseConfig(
           cliOptionsOrConfig.configFile,
           cliOptionsOrConfig,
           {
             promiseConfig: false,
             throwErrors: true
           }
         )
       } catch (parseConfigError) {
         // TODO: change how `done` falls back to exit in next major version
         //  SEE: https://github.com/karma-runner/karma/pull/3635#discussion_r565399378
         done(1)
         return
       }
     }
```

Reply:
Exactly!